core/domain/change_domain.py

```diff
@@ -23,10 +23,20 @@
 from core import utils
 from core.platform import models
 
+from typing import Any, Dict, List
+
+MYPY = False
+if MYPY:  # pragma: no cover
+    from mypy_imports import base_models
+
 (base_models,) = models.Registry.import_models([models.NAMES.base_model])
 
 
-def validate_cmd(cmd_name, valid_cmd_attribute_specs, actual_cmd_attributes):
+def validate_cmd(
+    cmd_name: str, 
+    valid_cmd_attribute_specs: Dict[str, Any],
```

Comment:
Can this be stricter?

Reply:
I tried all the possibilities. Ideally it should it return `List[str]`, as `required_attribute_names` is a list of required attribute names of a command. But it Mypy is throwing the following errors 
![image](https://user-images.githubusercontent.com/65958420/148172529-51ed5320-c199-445a-a59b-7093c6f534f7.png)
Any suggestions?

Reply:
@vojtechjelinek This is the error that I am getting if I use `Dict[str,<something>]`

![image](https://user-images.githubusercontent.com/65958420/148416337-a48c5fbc-9cce-462b-8090-fefebb3aedc2.png)

And the reason why I think it should return List[str] is `required_attribute_names = valid_cmd_attribute_specs['required_attribute_names]`, which means that type annotation of `required_attribute_names` should be the return type of `valid_cmd_attribute_specs`, and in the comments in the later lines 'required_attribute_names (a list of required attribute names of a command)', required_attribute_names is defined as this.  It is the reason, I think it should be List[str]. So, I don't think it is possible to make the type annotation stricter, in this scenario :(

Reply:
The `valid_cmd_attribute_specs` has some predefined structure, you can see example of that on a screenshot below (except for the `name` as that is removed on line 202)
![image](https://user-images.githubusercontent.com/10142938/148687617-ce3f10d4-4994-4c4b-bf30-e58c0e785844.png). Can we use `TypedDict` to create a type that will (at least partially) conform to that structure?

Reply:
@vojtechjelinek , I tried using `TypedDict` for this purpose, as specified in the code above

![image](https://user-images.githubusercontent.com/65958420/148801785-69b6dc32-ddc0-4383-b3b6-1dc34420d051.png)

and on Mypy is throwing below error

![image](https://user-images.githubusercontent.com/65958420/148802072-5ae85b7f-c1b5-4665-a205-7bd5c23859ea.png)

which is because of the lines, https://github.com/div-yam/oppia/blob/9d78333631fc8b680d307527c28b757e4aca5968/core/domain/change_domain.py#L207-L208, which in my opinion is due to https://github.com/div-yam/oppia/blob/9d78333631fc8b680d307527c28b757e4aca5968/core/domain/change_domain.py#L121 . Now I am not able to comprehend what to do in this case.

Reply:
Then the type of `ALLOWED_COMMANDS` should be `List[valid_cmd_attribute_specs]`, right?

Reply:
After using this command, I am getting the following errors
```
(oppia) dbm@dbm-HP-Notebook:~/gsoc/oppia$ python -m scripts.run_mypy_checks  --skip-install
Installing Mypy and stubs for third party libraries.
Installed Mypy and stubs for third party libraries.
Starting Mypy type checks.
core/domain/change_domain.py:161: error: Unsupported operand types for + ("List[valid_cmd_attribute_specs]" and "List[Dict[str, Sequence[str]]]")  [operator]
core/domain/change_domain.py:197: error: Unsupported operand types for + ("List[valid_cmd_attribute_specs]" and "List[Dict[str, Sequence[str]]]")  [operator]
core/domain/change_domain.py:210: error: Key 'name' of TypedDict "valid_cmd_attribute_specs" cannot be deleted  [misc]
core/domain/change_domain.py:228: error: Unsupported operand types for + ("List[valid_cmd_attribute_specs]" and "List[Dict[str, Sequence[str]]]")  [operator]
core/domain/rights_domain.py:321: error: Incompatible types in assignment (expression has type "List[Dict[str, Union[str, List[str], Dict[str, Union[str, List[str]]]]]]", base class "BaseChange" defined the type as "List[valid_cmd_attribute_specs]")  [assignment]
core/domain/rights_domain.py:340: error: Incompatible types in assignment (expression has type "List[Dict[str, Union[str, List[str], Dict[str, Union[str, List[str]]]]]]", base class "BaseChange" defined the type as "List[valid_cmd_attribute_specs]")  [assignment]
Found 6 errors in 2 files (checked 218 source files)
```
Sorry, for some reason I am not able to click screenshots on my system 😅
Here Line 161 refers to line 153, Line 197 refers to line 189, Line 210 refers to line 202, and Line 228 refers to Line 220 in https://github.com/div-yam/oppia/blob/a7fe592852252187c286a87975cc9c0b32208eb9/core/domain/change_domain.py

Reply:
Okay, please try to investigate the errors more, do not just expect the answer from me. I think here you need to also type the `COMMON_ALLOWED_COMMANDS` as `List[valid_cmd_attribute_specs]` and it might help. If not you need to investigate how to merge the dictionaries in a different way.

Reply:
@vojtechjelinek 
I have been on figuring out for a long time, how to go about this issue. Why can't just we use `valid_cmd_attribute_specs` annotation as `Dict[str, List[str]]` directly. I am getting the following error.

![image](https://user-images.githubusercontent.com/65958420/149806210-b787c77f-de8e-430d-ae7e-2d88df6be0bc.png)

To fix the first two errors I used `Any` type annotation for the `deprecated_values`, I got the following error.

![image](https://user-images.githubusercontent.com/65958420/149806541-0c98483f-c5fa-4b78-bc74-743e3deb86f4.png)

Reply:
Ah I see one problem, `valid_cmd_attribute_specs` is both name of a variable bit also a name of type, change the type name to `ValidCmdAttributeSpecsDict` and elts see where we get.

Reply:
@vojtechjelinek , I changed the type name to `ValidCmdAttributeSpecsDict`, and I am getting the same errors as above.

![image](https://user-images.githubusercontent.com/65958420/150587575-c9da55dd-4263-4442-83df-ee7705a1016d.png)

Reply:
How do you type `valid_cmd_attribute_specs` in the `validate_dict` function? I think it should also have the same type (`ValidCmdAttributeSpecsDict`), then you will probably get other errors, linked to the type of `ALLOWED_COMMANDS` which should also probably be `List[ValidCmdAttributeSpecsDict]`. Maybe you also need to make `ValidCmdAttributeSpecsDict` not total. You can find how in the MyPy documentation.

Reply:
I tried to make `ValidCmdAttributeSpecsDict` 's Totality false, by the following the code

![image](https://user-images.githubusercontent.com/65958420/150640590-088099bd-2aa0-4db0-b4ac-0481b28ae257.png)

and it threw the following errors

![image](https://user-images.githubusercontent.com/65958420/150640607-f35580dd-1bbd-4d49-a162-a47b4495297e.png)